src/tests/ftest/util/apricot/apricot/test.py

```diff
@@ -124,61 +129,95 @@ def setUp(self):
         self.d_log = DaosLog(self.context)
 
     def tearDown(self):
+        """Tear down after each test case."""
         super(TestWithoutServers, self).tearDown()
 
         if self.fault_file:
             os.remove(self.fault_file)
 
+
 class TestWithServers(TestWithoutServers):
-    '''
-    Run tests with DAOS servers
+    """Run tests with DAOS servers and at least one client.
+
+    Optionally run DAOS clients on specified hosts.  By default run a single
+    DAOS client on the host executing the test.
 
     :avocado: recursive
-    '''
+    """
+
     def __init__(self, *args, **kwargs):
+        """Initialize a TestWithServers object."""
         super(TestWithServers, self).__init__(*args, **kwargs)
 
         self.agent_sessions = None
 
     def setUp(self):
+        """Set up each test case."""
         super(TestWithServers, self).setUp()
 
-        self.server_group = self.params.get("server_group", '/server/',
-                                            'daos_server')
-        self.hostlist_servers = self.params.get("test_machines", '/run/hosts/*')
-
+        self.server_group = self.params.get(
+            "server_group", "/server/", "daos_server")
+
+        # Determine which hosts to use as servers and optionally clients.
+        # Support the use of a host type count to test with subsets of the
+        # specified hosts lists
+        test_machines = self.params.get("test_machines", "/run/hosts/*")
+        test_servers = self.params.get("test_servers", "/run/hosts/*")
+        test_clients = self.params.get("test_clients", "/run/hosts/*")
+        server_count = self.params.get("server_count", "/run/hosts/*")
+        client_count = self.params.get("client_count", "/run/hosts/*")
+
+        # Supported combinations of yaml hosts arguments:
+        #   - test_machines [+ server_count]
+        #   - test_servers [+ server_count]
+        #   - test_servers [+ server_count] + test_clients [+ client_count]
+        if test_machines:
+            self.hostlist_servers = test_machines[:server_count]
```

Comment:
Doesn't this imply that every test's `.yaml` file has to define a `server_count` when this patch lands?  If so, I'd propose that counts default to the total count of nodes provided in the `test_{machines,servers,clients}` lists so that tests don't need to specify a count when they want all nodes.

(Ditto for the same uses below).

Reply:
The `server_count` and `client_count `yaml file entries are not required.  If they do not exist in the yaml file they will default to `None` and `list[:None]` is the same as `list`.

Reply:
Ahhh.  Interesting.  I don't think I recall ever seeing an list indexing value of `None` being used before.